### DISCUSSION TITLE
New daily rate limit on /statuses/mentions and /statuses/user_timeline

### DISCUSSION THREAD

**Discussion by ryanjgallagher** [tweepy/tweepy#1240] [Questions] [answered]
On June 19, Twitter introduced 24-hour rate limits on the `/statuses/mentions` and `/statuses/user_timeline` endpoints that specify an application can't make more than 100,000 requests total in a 24-hour period: https://developer.twitter.com/en/docs/tweets/timelines/FAQ.html

Does Tweepy's current handling of rate limiting handle this new type of rate limit, or does it need to be modified to do so?

If it doesn't need to be modified, it may still be nice to mention this different type of rate limit in the "Code Snippets" section of the documentation that mentions the 15-min rate limits.

I tried looking through the commits and issues to see if this has brought up yet, and I didn't see anything, but I apologize if this is a duplicate.

**Comment by Harmon758** [comment, marked as answer]
The 429 error response returned when the 100,000 request limit is exceeded in a 24-hour period is the same as any other endpoint exceeding its normal rate limit.
See https://developer.twitter.com/en/docs/basics/rate-limiting.

So if `wait_on_rate_limit` is set to `False`, it should raise `RateLimitError` as expected, and it will be up to the user to handle it properly.

If `wait_on_rate_limit` is set to `True`, the library will attempt to handle it as a normal 15-min rate limit. The workaround for this would be to specify a custom `retry_delay` specifically when using these endpoints.

The problem with adding additional handling for these new rate limits is that, per [the announcement](https://twittercommunity.com/t/mentions-and-user-timeline-request-limit-goes-into-effect-today/127044):
> The existing `x-rate-limit-*` headers only apply to the 15-minute window rate limits. At this point, we are not passing additional headers in the response around the new 100,000 request limit.

and per the FAQ you linked:
> How do I check whether I am hitting the endpoint(s) near the request limit?
> 
> We do not have an endpoint that provides this information, so you will have to review your logs to track your usage. If you would like to check your usage from the current rate limit window, you can use the [application/rate_limit_status](https://developer.twitter.com/en/docs/developer-utilities/rate-limit-status/api-reference/get-application-rate_limit_status) endpoint.

The current rate limit window being referred to is the 15-minute window and the application/rate_limit_status endpoint only provides information about that window.

This means that there's no way to track or even check directly from the API how much of the rate limit has been used or when to retry when hitting the rate limit.
I don't think it's within the scope of the library to hardcode the rate limits and internally track how many requests you've made to those endpoints and when you made the first request (since from the FAQ, a rolling clock is used for the 24-hour period).
Also, from [this forum post answer](https://twittercommunity.com/t/question-about-rate-limit-changes-to-mentions-timeline-and-user-timeline/126958/2) by Twitter staff, it seems like they might add additional header information for the new 24-hour request limit in the future, which the library should be able to handle.

As for the documentation, it should always be supplemented by Twitter's API documentation, and the documentation for the [statuses/user_timeline](https://developer.twitter.com/en/docs/tweets/timelines/api-reference/get-statuses-user_timeline.html) and [statuses/mentions_timeline](https://developer.twitter.com/en/docs/tweets/timelines/api-reference/get-statuses-mentions_timeline.html) endpoints specifically cover the new rate limits.
Also, currently, specific rate limits aren't covered in Tweepy's documentation.
Regarding the ["Handling the rate limit using cursors" subsection of the "Code Snippets" section of Tweepy's documentation](https://tweepy.readthedocs.io/en/latest/code_snippet.html#handling-the-rate-limit-using-cursors), the part that says "wait for 15 minutes" is referring to the code snippet shown. The subsection itself is specifically about how to handle rate limits when using cursors, and I don't think information about rate limits, either the standard 15-min ones or the new 24-hour ones, belongs in a section of the documentation dedicated to example code snippets.

If you're encountering any unexpected behavior/issues with the new rate limits, feel free to open a new issue, or if you'd like to add a new section in the documentation that covers rate limits, PRs are welcome.
